src/kernel.c

```diff
@@ -0,0 +1,640 @@
+	// the first thread in the ready list. There is always at least a thread
+	// here: the idle thread.
+	th_cur = __thlist_rem(&th_ready);
+	assert(th_cur != NULL);
+	debugf("[kernel] switching to %s(%p) PC=%lx\n", th_cur->name, th_cur, th_cur->stack_state->epc);
+	assert(!(th_cur->flags & TH_FLAG_INLIST));
+
+	// Set the current interrupt depth to that of the current thread.
+	__interrupt_depth = th_cur->interrupt_depth;
+
+	return th_cur->stack_state;
+}
+
+/**
+ * @brief Idle thread function
+ *
+ * Since there must always be something run by the CPU, the kernel creates
+ * a "idle" thread: a lowest-priority thread that is run only when all
+ * user-created threads are blocked.
+ *
+ * This thread does absolutely nothing. It just waits for any other thread
+ * to wake up so that it will be preempted.
+ */
+void __kthread_idle(void *arg)
+{
+	// When this thread is scheduled, there are no
+	// ready threads and all threads are probably waiting for some
+	// hardware event to happen. Just spin-wait here.
+	while(1) {}
+}
+
+kthread_t* kernel_init(void)
+{
+	assert(!__kernel);
+	th_ready = NULL;  // empty ready list
+	th_count = 1; // start with the main thread
+
+	// Configure the main thread
+	memset(&th_main, 0, sizeof(th_main));
+	th_main.pri = 0;
+	strlcpy(th_main.name, "main", sizeof(th_main.name));
+
+	// NOTE: keep this in sync with system.c
+	#define STACK_SIZE 0x10000
+	th_main.stack = (char*)0x80000000 + get_memory_size() - STACK_SIZE;
+
+	uint64_t *s = (uint64_t*)th_main.stack;
+	for (int i=0;i<STACK_GUARD/8;i++)
+		s[i] = STACK_COOKIE;
+
+	// The main thread is the currently scheduled one.
+	th_cur = &th_main;
+
+	// Allocate the idle thread
+	th_idle = kthread_new("idle", 4096, -1, __kthread_idle, 0);
+
+	__kernel = true;
+	return th_cur;
+}
+
+void kernel_close(void)
+{
+	assert(__kernel);
+	assertf(th_cur == &th_main, "kthread_close can only be called from main thread");
+
+	// Kill the idle thread to release the memory
+	kthread_kill(th_idle);
+	th_idle = NULL;
+
+	assertf(th_count == 1, "not all threads were killed");
+
+	th_cur = NULL;
+	__kernel = false;
+	__isr_force_schedule = false;
+}
+
+kthread_t* kthread_new(const char *name, int stack_size, int pri, void (*user_entry)(void*), void *user_data)
+{
+	assert((stack_size % 8) == 0);
+
+	// Allocate memory for the stack, the stack guard area, and the kthread_t
+	// structure
+	//
+	// |             |                        |             |
+	// | Stack guard |   Stack                | kthread_t   |
+	// |             |                        |             |
+	//
+	//
+	// For the kthread structure, we will use the final part of
+	// the buffer: since the stack grows downward, this makes sure that
+	// overflowing the stack doesn't corrupt the thread structure, which might
+	// make debugging more difficult and might even cause the kernel to crash.
+	void *thmem = malloc(STACK_GUARD + stack_size + sizeof(kthread_t));
```

Comment:
I'd be curious to hear your thoughts on returning `NULL` if `malloc` returns `NULL`.

I worry it might make the kernel API more complicated, but I can also see homebrew programmers coming from a wide background with varying familiarity of embedded programming. 

An assertion might be a better middle ground.

Reply:
I don't like returning NULL here because the return error can safely be ignored most of the times. In generale, I don't expect N64 games to gracefully handle dynamic end-of-memory situations, so I will just add an assert here to simplify debugging.